content/articles/2017-09-erratum-tidyr.Rmd

```diff
@@ -0,0 +1,207 @@
+that version, context expressions cannot refer to the data. This makes
+it safer to refer to other objects.
+
+
+### Safety versus convenience
+
+Most datawise functions in R obey the same set rules regarding the
+visibility of objects. Visibility is hierarchical: data frame objects
+override those found in the context. This is convenient for
+interactive use and scripts, but can also cause issues. In the
+following example, should `gather()` select the first three columns
+(using the `x` defined in the global environment), or should it select
+the first two columns (using the column named `x`)?
+
+```{r}
+x <- 3
+df <- tibble(w = 1, x = 2, y = 3)
+df %>% gather("key", "value", 1:x)
+```
+
+The answer is that it selects the first two variables because `x` is
+first found in the data.
+
+In practice, the hierarchical ambiguity is not a big problem when you
+use these tools interactively or in scripts. However it becomes
+worrying when you're writing reusable functions because you don't know
+in advance the variables in the input data frame. For those cases
+where ambiguity matters, the tidy eval feature of quasiquotation
+allows you to explicitly pick a variable from the context:
+
+```{r}
+df %>% gather("key", "value", 1 : (!! x))
+```
+
+
+### The special semantics of selection functions
+
+Selection functions have always been a bit special in the tidyverse.
+They don't behave exactly like regular quoting functions. In almost
+all quoting functions in R, variables represent a data frame *column*.
+That's why expressions like this are natural:
+
+```{r, eval = FALSE}
+lm(mass ~ scale(height), data = starwars)
+```
+
+Since variables represent actual columns, you can include them in
+expressions as if they were actual objects. For instance
+`scale(height)` is equivalent to `scale(starwars$height)`. The same is
+true for most tidyverse tools, e.g. `dplyr::mutate()`:
+
+```{r, eval = FALSE}
+starwars %>% mutate(height = scale(height))
+```
+
+However in selection functions, variables do not represent columns but
+*column positions*. That is a subtle but important distinction. When
+you type `height`, tidyselect actually sees the integer `2`. This
+makes sense for several reasons:
+
+* Expressions such as `name:mass` evaluates to `1:3` in a natural way.
+
+* You can select the same column multiple times. For example if you
+  supply a selection for the dataset `starwars` with
+  `starts_with("s")` and `ends_with("s")`, the variables `species` and
+  `starships` would be matched twice. It is easy for tidyselect to
+  take the intersection of the two sets of column positions. If the
+  sets contained vectors instead, it could not determine whether it
+  got two different but identical vectors rather than the same
+  vector selected twice.
+
+* Finally and most importantly, if the variables evaluated to the
+  column vectors, we would have no information about their names or
+  positions, which we need to reconstitute the data frame.
+
+
```

Comment:
Should we mention that this allows you to do silly things like `select(starwars, sqrt(hair_color):mass^2)`?  (But we expect that you won't do that)